File: service_detector.py
"""Service detection from SMS data"""

SERVICE_MAP = {
    'google': 'Google', 'gmail': 'Google', 'g.co': 'Google',
    'whatsapp': 'WhatsApp', 'wa': 'WhatsApp',
    'telegram': 'Telegram', 'tg': 'Telegram',
    'facebook': 'Facebook', 'fb': 'Facebook', 'meta': 'Facebook',
    'instagram': 'Instagram', 'ig': 'Instagram',
    'twitter': 'Twitter', 'x.com': 'Twitter',
    'tiktok': 'TikTok',
    'uber': 'Uber',
    'amazon': 'Amazon',
    'netflix': 'Netflix',
    'spotify': 'Spotify',
    'linkedin': 'LinkedIn',
    'snapchat': 'Snapchat',
    'discord': 'Discord',
    'microsoft': 'Microsoft', 'outlook': 'Microsoft', 'hotmail': 'Microsoft',
    'apple': 'Apple', 'icloud': 'Apple',
    'yahoo': 'Yahoo',
    'paypal': 'PayPal',
    'stripe': 'Stripe',
    'openai': 'OpenAI',
    'foodpanda': 'Foodpanda', 'panda': 'Foodpanda',
    'grab': 'Grab',
    'line': 'LINE',
    'wechat': 'WeChat',
    'viber': 'Viber',
    'signal': 'Signal',
    'kakao': 'Kakao',
    'binance': 'Binance',
    'coinbase': 'Coinbase',
    'steam': 'Steam',
    'tinder': 'Tinder',
    'venmo': 'Venmo',
    'cashapp': 'CashApp',
    'protonmail': 'ProtonMail',
    'tutanota': 'Tutanota',
    'yandex': 'Yandex',
    'vk': 'VK',
    'shein': 'SHEIN',
    'shopee': 'Shopee',
    'paytm': 'Paytm',
    'gpay': 'Google Pay',
    'samsung': 'Samsung',
    'xiaomi': 'Xiaomi',
    'oneplus': 'OnePlus',
    'oppo': 'OPPO',
    'vivo': 'Vivo',
    'huawei': 'Huawei',
    'realme': 'Realme',
}

import re

MESSAGE_PATTERNS = [
    (re.compile(r'\bwhatsapp\b', re.I), 'WhatsApp'),
    (re.compile(r'\btelegram\b', re.I), 'Telegram'),
    (re.compile(r'\bfacebook\b|\bmeta\b', re.I), 'Facebook'),
    (re.compile(r'\binstagram\b', re.I), 'Instagram'),
    (re.compile(r'\btwitter\b|\bx\.com\b', re.I), 'Twitter'),
    (re.compile(r'\btiktok\b', re.I), 'TikTok'),
    (re.compile(r'\bgoogle\b|\bgmail\b', re.I), 'Google'),
    (re.compile(r'\bmicrosoft\b|\boutlook\b|\bhotmail\b', re.I), 'Microsoft'),
    (re.compile(r'\bapple\b|\bicloud\b', re.I), 'Apple'),
    (re.compile(r'\bamazon\b', re.I), 'Amazon'),
    (re.compile(r'\bnetflix\b', re.I), 'Netflix'),
    (re.compile(r'\bspotify\b', re.I), 'Spotify'),
    (re.compile(r'\blinkedin\b', re.I), 'LinkedIn'),
    (re.compile(r'\bsnapchat\b', re.I), 'Snapchat'),
    (re.compile(r'\bdiscord\b', re.I), 'Discord'),
    (re.compile(r'\bpaypal\b', re.I), 'PayPal'),
    (re.compile(r'\bstripe\b', re.I), 'Stripe'),
    (re.compile(r'\buber\b', re.I), 'Uber'),
    (re.compile(r'\byahoo\b', re.I), 'Yahoo'),
    (re.compile(r'\bline\b', re.I), 'LINE'),
    (re.compile(r'\bviber\b', re.I), 'Viber'),
    (re.compile(r'\bsignal\b', re.I), 'Signal'),
    (re.compile(r'\bgrab\b', re.I), 'Grab'),
    (re.compile(r'\bopenai\b', re.I), 'OpenAI'),
    (re.compile(r'\bfoodpanda\b', re.I), 'Foodpanda'),
    (re.compile(r'\bcoinbase\b', re.I), 'Coinbase'),
    (re.compile(r'\bbinance\b', re.I), 'Binance'),
    (re.compile(r'\bsteam\b', re.I), 'Steam'),
    (re.compile(r'\btinder\b', re.I), 'Tinder'),
    (re.compile(r'\bshein\b', re.I), 'SHEIN'),
    (re.compile(r'\bshopee\b', re.I), 'Shopee'),
    (re.compile(r'\bpaytm\b', re.I), 'Paytm'),
]

def _normalize_service_name(raw: str) -> str | None:
    if not raw or not isinstance(raw, str):
        return None
    cleaned = raw.strip().lower()
    if not cleaned:
        return None
    if cleaned in SERVICE_MAP:
        return SERVICE_MAP[cleaned]
    for key, canonical in SERVICE_MAP.items():
        if key in cleaned:
            return canonical
    return None

def detect_service(from_field: str = None, service_field: str = None, message: str = None) -> str | None:
    # Priority 1: from field (sender name = service)
    if from_field:
        result = _normalize_service_name(from_field)
        if result:
            return result
    
    # Priority 2: explicit service field
    if service_field:
        result = _normalize_service_name(service_field)
        if result:
            return result
    
    # Priority 3: keyword matching in message
    if message and isinstance(message, str):
        for pattern, service in MESSAGE_PATTERNS:
            if pattern.search(message):
                return service
    
    return None

File: src/lib/service-detector.ts
/**
 * Service detection from SMS data
 * Priority: from field → service field → message keywords
 */

/**
 * Canonical service name mapping (case-insensitive)
 */
export const SERVICE_MAP: Record<string, string> = {
  'google': 'Google',
  'gmail': 'Google',
  'g.co': 'Google',
  'whatsapp': 'WhatsApp',
  'wa': 'WhatsApp',
  'telegram': 'Telegram',
  'tg': 'Telegram',
  'facebook': 'Facebook',
  'fb': 'Facebook',
  'meta': 'Facebook',
  'instagram': 'Instagram',
  'ig': 'Instagram',
  'twitter': 'Twitter',
  'x.com': 'Twitter',
  'x': 'Twitter',
  'tiktok': 'TikTok',
  'uber': 'Uber',
  'amazon': 'Amazon',
  'netflix': 'Netflix',
  'spotify': 'Spotify',
  'linkedin': 'LinkedIn',
  'snapchat': 'Snapchat',
  'discord': 'Discord',
  'microsoft': 'Microsoft',
  'outlook': 'Microsoft',
  'hotmail': 'Microsoft',
  'live': 'Microsoft',
  'apple': 'Apple',
  'icloud': 'Apple',
  'yahoo': 'Yahoo',
  'paypal': 'PayPal',
  'stripe': 'Stripe',
  'openai': 'OpenAI',
  'foodpanda': 'Foodpanda',
  'panda': 'Foodpanda',
  'grab': 'Grab',
  'line': 'LINE',
  'wechat': 'WeChat',
  'viber': 'Viber',
  'signal': 'Signal',
  'kakao': 'Kakao',
  'naver': 'Naver',
  'paypal': 'PayPal',
  'venmo': 'Venmo',
  'cashapp': 'CashApp',
  'coinbase': 'Coinbase',
  'binance': 'Binance',
  'steam': 'Steam',
  'epic': 'Epic Games',
  'riot': 'Riot Games',
  'ea': 'EA',
  'origin': 'EA',
  'ubisoft': 'Ubisoft',
  'yahoo': 'Yahoo',
  'protonmail': 'ProtonMail',
  'tutanota': 'Tutanota',
  'mailru': 'Mail.ru',
  'yandex': 'Yandex',
  'vk': 'VK',
  'ok': 'OK',
  'tinder': 'Tinder',
  'badoo': 'Badoo',
  'bumble': 'Bumble',
  'hinge': 'Hinge',
  'doordash': 'DoorDash',
  'lyft': 'Lyft',
  'ola': 'Ola',
  'mercadolibre': 'MercadoLibre',
  'mercado': 'MercadoLibre',
  'alibaba': 'Alibaba',
  'aliexpress': 'AliExpress',
  'shein': 'SHEIN',
  'shopee': 'Shopee',
  'lazada': 'Lazada',
  'tokopedia': 'Tokopedia',
  'bukalapak': 'Bukalapak',
  'blibli': 'Blibli',
  'ola': 'Ola',
  'swiggy': 'Swiggy',
  'zomato': 'Zomato',
  'deliveroo': 'Deliveroo',
  'justeat': 'Just Eat',
  'paytm': 'Paytm',
  'phonepe': 'PhonePe',
  'gpay': 'Google Pay',
  'samsung': 'Samsung',
  'huawei': 'Huawei',
  'xiaomi': 'Xiaomi',
  'oppo': 'OPPO',
  'vivo': 'Vivo',
  'oneplus': 'OnePlus',
  'realme': 'Realme',
};

/**
 * Message keyword patterns for service detection
 * Ordered by specificity (more specific first)
 */
const MESSAGE_PATTERNS: Array<{ pattern: RegExp; service: string }> = [
  { pattern: /\bwhatsapp\b/i, service: 'WhatsApp' },
  { pattern: /\btelegram\b/i, service: 'Telegram' },
  { pattern: /\bfacebook\b|\bmeta\b/i, service: 'Facebook' },
  { pattern: /\binstagram\b/i, service: 'Instagram' },
  { pattern: /\btwitter\b|\bx\.com\b/i, service: 'Twitter' },
  { pattern: /\btiktok\b/i, service: 'TikTok' },
  { pattern: /\bgoogle\b|\bgoogle\b/i, service: 'Google' },
  { pattern: /\bgmail\b/i, service: 'Google' },
  { pattern: /\bmicrosoft\b|\boutlook\b|\bhotmail\b/i, service: 'Microsoft' },
  { pattern: /\bapple\b|\bicloud\b/i, service: 'Apple' },
  { pattern: /\bamazon\b/i, service: 'Amazon' },
  { pattern: /\bnetflix\b/i, service: 'Netflix' },
  { pattern: /\bspotify\b/i, service: 'Spotify' },
  { pattern: /\blinkedin\b/i, service: 'LinkedIn' },
  { pattern: /\bsnapchat\b/i, service: 'Snapchat' },
  { pattern: /\bdiscord\b/i, service: 'Discord' },
  { pattern: /\bpaypal\b/i, service: 'PayPal' },
  { pattern: /\bstripe\b/i, service: 'Stripe' },
  { pattern: /\buber\b/i, service: 'Uber' },
  { pattern: /\byahoo\b/i, service: 'Yahoo' },
  { pattern: /\bline\b/i, service: 'LINE' },
  { pattern: /\bviber\b/i, service: 'Viber' },
  { pattern: /\bsignal\b/i, service: 'Signal' },
  { pattern: /\bgrab\b/i, service: 'Grab' },
  { pattern: /\bopenai\b/i, service: 'OpenAI' },
  { pattern: /\bfoodpanda\b/i, service: 'Foodpanda' },
  { pattern: /\bcoinbase\b/i, service: 'Coinbase' },
  { pattern: /\bbinance\b/i, service: 'Binance' },
  { pattern: /\bsteam\b/i, service: 'Steam' },
  { pattern: /\btinder\b/i, service: 'Tinder' },
  { pattern: /\blyft\b/i, service: 'Lyft' },
  { pattern: /\bdoordash\b/i, service: 'DoorDash' },
  { pattern: /\bshein\b/i, service: 'SHEIN' },
  { pattern: /\bshopee\b/i, service: 'Shopee' },
  { pattern: /\bpaytm\b/i, service: 'Paytm' },
];

/**
 * Normalize a service name using the canonical mapping
 */
function normalizeServiceName(raw: string): string | null {
  if (!raw || typeof raw !== 'string') return null;

  const cleaned = raw.trim().toLowerCase();
  if (!cleaned) return null;

  // Direct match in SERVICE_MAP
  if (SERVICE_MAP[cleaned]) {
    return SERVICE_MAP[cleaned];
  }

  // Try matching as a key within the raw string
  for (const [key, canonical] of Object.entries(SERVICE_MAP)) {
    if (cleaned.includes(key)) {
      return canonical;
    }
  }

  return null;
}

/**
 * Detect service from available fields
 * Priority: from field → service field → message keywords
 */
export function detectService(
  fromField?: string,
  serviceField?: string,
  message?: string
): string | null {
  // Priority 1: from field (sender name = service)
  if (fromField) {
    const fromService = normalizeServiceName(fromField);
    if (fromService) return fromService;
  }

  // Priority 2: explicit service field in payload
  if (serviceField) {
    const serviceFromField = normalizeServiceName(serviceField);
    if (serviceFromField) return serviceFromField;
  }

  // Priority 3: Keyword matching in message text
  if (message && typeof message === 'string') {
    for (const { pattern, service } of MESSAGE_PATTERNS) {
      if (pattern.test(message)) {
        return service;
      }
    }
  }

  return null;
}
